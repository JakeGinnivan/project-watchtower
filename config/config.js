/*
 * for testing only
 */
exports.default = {
    LINT_EXCLUDE: [
        'demo/**'
    ]
}
