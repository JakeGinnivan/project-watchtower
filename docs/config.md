# Configuration and Structure

## Default project structure

See the [build documentation](./build.md) for details on how to customise the project structure.

### Source locations

* config (optional. TypeScript has to be transpiled to JavaScript before building!)
    * webpack.[target].[environment].js: Override configuration
    * webpack-hooks.js: Add webpack options
    * config.js: Override default configuration
    * setup-tests.ts/js: Jest test setup
* common: Application code
* client
    * index.tsx: Client entry file
    * polyfills.ts: Polyfills (optional)
* public: Public directory for assets that aren't bundled with the application
    * index.html: HTML template for client-only mode
* server
    * start.ts: Server entry file
* .env: Environment variables
* .env.default: Definition of environment variables

### Output locations

* build/client: Client build output directory
    * static
        * js
            * main.js (hashed in production build)
            * vendor.js (hashed in production build)
        * css
            * main.css (hashed in production build)
        * media
        * fonts
* build/server: Server build output directory
    * *server.js*: Main file generated for server
    * ...
    * index.html
* assets.json: Mapping to the location of the generated assets
* build-stats.csv: Generated build metrics

## Override default configuration

Options for the file  `/config/config.js` (usage explained in the [build documentation](./build.md))

```ts
export interface BuildConfig {

    /** Prefix for all assets (JS, CSS, media, fonts) with trailing slash */
    ASSETS_PATH_PREFIX: string

    /** root path of your application */
    BASE: string

    /** entry file for the client */
    CLIENT_ENTRY: string

    /** file for client polyfills if needed */
    CLIENT_POLYFILLS: string | false

    /** output directory for the client build */
    CLIENT_OUTPUT: string

    /** Autoprefixer browser compatibilty  */
    CSS_AUTOPREFIXER: string[]

    /** set to false if the application is serverless */
    HAS_SERVER: boolean,

    /** List paths to exclude from linting */
    LINT_EXCLUDE: string[]

    /** Paths where modules are resolved */
    MODULE_PATHS: string[]

    /** Default port for the server (when process.env.PORT is not set) */
    PORT: number

    /** URL prefix for all resources */
    PUBLIC_PATH: string

    /** entry file for the server if applicable */
    SERVER_ENTRY: string

    /** if true, all externals will be bundled */
    SERVER_BUNDLE_ALL: boolean

    /** modules which the server build includes in the bundle */
    SERVER_INCLUDE_IN_BUNDLE: string[]

    /** output directory of the server */
    SERVER_OUTPUT: string

    /** directory that is served as static resources */
    SERVER_PUBLIC_DIR: string | false

    /** if true, no hash is added to the generated assets */
    STATIC_RESOURCE_NAMES: boolean

    /** Additional environment variables for build stats */
    STATS_ENV: { [key: string]: string }

    /** Pages to run build stats on, format { name: URL } */
    STATS_PAGES: { [name: string]: string }

    /** Regular expression of paths to be ignored in watch mode */
    WATCH_IGNORE: RegExp

}
```

Default configuration:

```ts
{
    ASSETS_PATH_PREFIX: 'static/',
    ASSETS_ROOT: path.resolve(root),
    BASE: root,
    CLIENT_ENTRY: path.resolve(root, 'client', 'index.tsx'),
    CLIENT_OUTPUT: path.resolve(root, 'dist'),
    CLIENT_POLYFILLS: path.resolve(root, 'client', 'polyfills.ts'),
    CSS_AUTOPREFIXER: ['last 2 versions'],
    HAS_SERVER: true,
    LINT_EXCLUDE: [],
    MODULE_PATHS: [
        root,
        path.resolve(root, 'node_modules'),
        path.resolve(root, 'common'),
        path.resolve(root, 'app'),
    ],
    PUBLIC_PATH: '/',
    PORT: 3000,
    /** if true, all externals will be bundled */
    SERVER_BUNDLE_ALL: false,
    SERVER_INCLUDE_IN_BUNDLE: [
        'project-watchtower',
    ],
    SERVER_ENTRY: path.resolve(root, 'server', 'start.ts'),
    SERVER_OUTPUT: path.resolve(root, 'dist'),
    SERVER_PUBLIC_DIR: path.resolve(root, 'public'),
    STATIC_RESOURCE_NAMES: false,
    STATS_ENV: {},
    STATS_PAGES: { home: '/' },
}
```

## Configuration file templates

Project Watchtower offers several templates for configuration files around tooling and building:

### TSLint: tslint.json

```json
{
    "extends": [
        "project-watchtower/presets/tslint"
    ]
}

```

### Jest: jest.config.js & jest.debug.config.js

If either of the above config files exist in the root they will be used

``` js
var baseConfig = require('project-watchtower/presets/jest/jest.json')

module.exports = Object.assign({}, baseConfig, {
    verbose: true,
    rootDir: '.',
    snapshotSerializers: ['enzyme-to-json/serializer'],
    moduleNameMapper: {
        '\\.(s?css|png|svg|jpg|eot|woff|woff2)$':
            '<rootDir>/node_modules/project-watchtower/lib/test/test-mapper.js',
        '^bundle-loader':
            '<rootDir>/node_modules/project-watchtower/lib/test/bundle-loader-mapper.js'
    },
    transform: {
        '.tsx?': '<rootDir>/node_modules/ts-jest/preprocessor.js'
    }
})

```

### SASS-Lint .sass-lint.yml

The sass-lint configuration can be referenced from `package.json`

```json
{
    ...
    "sasslintConfig": "node_modules/project-watchtower/presets/sass-lint/.sass-lint.yml"
}
```
