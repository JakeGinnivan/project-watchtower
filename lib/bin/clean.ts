import doClean from '../build/clean'
import PATHS from '../build/paths'

const { CLIENT_OUTPUT, SERVER_OUTPUT } = PATHS

const clean = (...paths: string[]) => (
    doClean([
        CLIENT_OUTPUT,
        SERVER_OUTPUT,
        '{client,common,config,server}/**/*.{js,map}',
        ...paths,
    ])
)

export default clean
