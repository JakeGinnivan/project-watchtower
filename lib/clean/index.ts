import * as rimraf from 'rimraf'
import { logError } from '../runtime/util/log'

const clean: (paths: string | string[]) => Promise<any> = paths => {
    if (Array.isArray(paths)) {
        return Promise.all(paths.map(path => clean(path)))
    }

    return new Promise((resolve, reject) => {
        rimraf(paths, err => {
            if (err) {
                logError('Clean error:', err)
                reject(err)
            } else {
                resolve()
            }
        })
    })
}

export default clean
