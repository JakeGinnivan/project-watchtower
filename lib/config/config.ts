import * as fs from 'fs'
import * as path from 'path'
import { logError } from '../util/log'
import { BuildConfig, BuildConfigOverride } from '../types'

const root = process.cwd()

const defaultConfig: BuildConfig = {
    ASSETS_PATH_PREFIX: 'static/',
    BASE: root,
    CLIENT_ENTRY: path.resolve(root, 'client', 'index.tsx'),
    CLIENT_OUTPUT: path.resolve(root, 'public', 'assets'),
    CLIENT_POLYFILLS: path.resolve(root, 'client', 'polyfills.ts'),
    CSS_AUTOPREFIXER: ['last 2 versions'],
    HAS_SERVER: true,
    LINT_EXCLUDE: [],
    MODULE_PATHS: [
        root,
        path.resolve(root, 'node_modules'),
        path.resolve(root, 'common'),
    ],
    PUBLIC_PATH: '/assets/',
    PORT: 3000,
    SERVER_ENTRY: path.resolve(root, 'server', 'start.ts'),
    SERVER_OUTPUT: path.resolve(root, 'build'),
    SERVER_PUBLIC_DIR: path.resolve(root, 'public'),
    STATIC_RESOURCE_NAMES: false,
    STATS_ENV: {},
    STATS_PAGES: { home: '/' },
    WATCH_IGNORE: /node_modules(?!.+swm-component-library)/,
}

const customConfigFile = path.resolve(root, 'config', 'config.js')
let customConfig: BuildConfigOverride = {}

try {
    if (fs.existsSync(customConfigFile)) {
        // tslint:disable-next-line no-var-requires
        customConfig = require(customConfigFile).default
    }
} catch (e) {
    logError('Error reading config/config.js!', e)
}

const CONFIG = {
    ...defaultConfig,
    ...customConfig,
}

export default CONFIG
