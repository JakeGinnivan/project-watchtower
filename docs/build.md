# Building, Running and Debugging

## Building

```
pwt build [<target> <environment>]
```

## Customising

By default, Project Watchtower uses its own webpack configuration files. If you want to modify or extend them with custom options for you project, there are multiple ways of providing custom configuration:

### Basic config file

If you want to change the default configuration of the build process, you can add a `/config/config.js` that overrides the settings used by the default webpack configuration:

```ts
import * as path from 'path'
import { BuildConfigOverride } from 'project-watchtower/lib/types'

const customConfig: BuildConfigOverride = {
    SERVER_OUTPUT: path.resolve(process.cwd(), 'dist')
}

export default customConfig
```

If you provide a `/config/config.ts` in TypeScript, make sure it is transpiled to JavaScript before any `pwt` command is run or any of Project Watchtower's middlewares are used. This can be done in the `prepare` script in your project's `package.json`. Project Watchtower will throw an error if it detects a TypeScript configuration file, but not the transpiled JavaScript artifact.

The complete default configuration is located in `project-watchtower/lib/runtime/config/config.ts`

See the [config documentation](./config.md) for an overview of the default project structure and configuration details.

### Webpack hooks

Add `/config/webpack-hooks.js` to your project:

```ts
import { WebpackHooks } from 'project-watchtower/lib/types'
import * as FooWebpackPlugin from 'foo-webpack-plugin'

const hooks: WebpackHooks = {
    base: {
        plugins: [
            new FooWebpackPugin()
        ]
    }
}

export default hooks
```

The following hooks can be defined:

* `base`
* `client`
* `server`
* `dev`
* `prod`
* `clientDev`
* `clientProd`
* `clientDebug`
* `serverDev`
* `serverProd`
* `serverDebug`

These hooks are applied internally to all webpack cofigurations that have a target and environment.

If you provide a completely custom webpack configuration as described below, they will
* not be applied if you only use basic building blocks (like `base`, `clientBase` or `devBase`)
* be applied if you extend complete configurations (like `clientDev`, `serverProd`)

### Custom webpack configuration

Add configuration files to your project: `/config/webpack.<target>.<environment>.js`.

Example:

```ts
import { merge } from 'project-watchtower/lib/build'
import { clientBase } from 'project-watchtower/lib/config'

const config = merge(
    base,
    clientBase,
    {
        // ...
    },
)

export default config
```

## Running

    pwt start [watch] [fast] [debug] [inspect] [prod]

Starts the server, using the environment variables defined in `.env`

### Environment Variables

*   `NODE_ENV`: set to `"production"` or `"development` depending on the `prod` flag - in the webpack bundles the values are hardcoded depending on which mode they were built in, so this is only relevant for third-party dependencies.
*   `START_WATCH_MODE`: set to `"true"` by the `watch` flag
*   `START_FAST_MODE`: set to `"true"` by the `fast` flag
*   `START_DEBUG_MODE`: set to `"true"` by the `debug` flag

If you want to use additional `process.env` variables in the **client** build, make sure you create a `.env` (for local) and `.env.default` file:

`.env`
> FOO=bar

`.env.default`
> FOO=

The values defined here are replaced in the client build. The server build still accesses its actual `process.env` object, only `process.env.NODE_ENV` is being replaced there.
Values that are present in the actual runtime environment at build time will _not_ be overridden by the ones defined in `.env` / `.env.default`. However, it is necessary to define all the environment variables in the `.env.default` file that are used in the client code.

### Hot Reloading

In development we use [webpack-dev-middleware](https://github.com/webpack/webpack-dev-middleware) and [webpack-hot-middleware](https://github.com/glenjamin/webpack-hot-middleware) to enable hot reloading. See the webpack docs under [https://webpack.js.org/guides/hot-module-replacement/](https://webpack.js.org/guides/hot-module-replacement/).

For CSS bundling we use the [extract-text-webpack-plugin](https://github.com/webpack-contrib/extract-text-webpack-plugin), as the usual development fallback to [style-loader](https://github.com/webpack-contrib/style-loader) seems to have problems referncing fonts through file-loader, and we want to stay as close to production as we can. As it does not support hot reloading out of the box yet, the project has to include a manual helper method:

```jsx
import { cssHotReload } from 'project-watchtower/lib/runtime/client'

const render = () => {
    const App = require<{ default: React.ReactType }>('./components/App').default
    ReactDOM.render((
        <BrowserRouter>
            <App />
        </BrowserRouter>
    ), document.getElementById('root'))
}

render()

if (module.hot) {
    module.hot.accept('./components/App', () => {
        setTimeout(render)
    })

    cssHotReload()
}
```

For more information see [https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/30#issuecomment-284301283](https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/30#issuecomment-284301283)

## Debugging

As VS code has problems with sourcemaps that are generated during the runtime of the debugger, the process of debugging has to be split into a "build" and a "debug" phase. The "build" tasks are defined in `.vscode/tasks.json`, the "debug" configurations in `.vscode/launch.json` - they reference the "build" tasks as their `preLaunchTask`.

### Debug server build

In your `package.json`, add a shortcut script for project-watchtower:

```
"pwt": "pwt"
```

Add a debug configuration to `.vscode/launch.json`:

```
{
    "type": "node",
    "request": "launch",
    "name": "Debug server",
    "runtimeExecutable": "npm",
    "windows": {
        "runtimeExecutable": "npm.cmd"
    },
    "runtimeArgs": [
        "run-script",
        "pwt",
        "start",
        "debug"
    ],
    "port": 5858,
    "timeout": 60000,
    "sourceMaps": true,
    "outFiles": [
        "${workspaceRoot}/build/server.js",
        "${workspaceRoot}/build/*.js"
    ],
    "smartStep": true,
    "trace": "sm",
    "preLaunchTask": "build:debug"
}
```

Add the following task to `.vscode/tasks.json`:

```
{
    "taskName": "build:debug",
    "args": [
        "run",
        "pwt",
        "build",
        "debug"
    ]
}
```

### Debug tests

Debugging test cases with Jest currently only works by transpiling the tests to JavaScript and then running them due to a limitation in the `ts-jest` preprocessor.

In your `package.json`, add a shortcut script for project-watchtower and the TypeScript compiler:

```
"pwt": "pwt",
"tsc": "tsc",
```

Add a debug configuration to `.vscode/launch.json`:

```
{
    "type": "node",
    "request": "launch",
    "name": "Debug tests",
    "runtimeExecutable": "npm",
    "windows": {
        "runtimeExecutable": "npm.cmd"
    },
    "runtimeArgs": [
        "run-script",
        "pwt",
        "test",
        "debug"
    ],
    "port": 5858,
    "timeout": 30000,
    "sourceMaps": true,
    "smartStep": true,
    "trace": "sm",
    "preLaunchTask": "build:test:debug"
}
```

Depending on your TypeScript configuration you might have to add the compilation target directory as `outFiles`:

```
"outFiles": [
    "${workspaceRoot}/dist/**/*.js"
]
```

Add a task to compile your tests to JavaScript to `.vscode/tasks.json`:

```
{
    "taskName": "build:test:debug",
    "args": [
        "run",
        "tsc",
        "--",
        "-p",
        "./tsconfig-debug.json"
    ]
}
```

### Server profiling

Start the server with

```
pwt start inspect
```

or

```
pwt watch inspect
```

Open Google Chrome and go to *chrome://inspect*, where the Node.js process will show up and allow you do to profiling.
