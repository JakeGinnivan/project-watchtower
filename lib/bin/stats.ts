import buildStats from '../stats'
import { BuildConfig } from '../../lib'

const stats = (buildConfig: BuildConfig, ...params: string[]) => {
    const verbose = params.indexOf('verbose') !== -1

    return buildStats(buildConfig, verbose)
}

export default stats
