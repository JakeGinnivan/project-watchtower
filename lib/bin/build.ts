import cleanBin from './clean'
import lint from './lint'
import test from './test'
import { ENVIRONMENTS, getWebpackConfig, TARGETS } from '../build/build'
import clean from '../clean'
import CONFIG from '../config/config'
import { failPromisesLate } from '../util/async'
import { webpackPromise } from '../util/webpack'
import { BuildEnvironment, BuildParam, BuildTarget } from '../types'

const { CLIENT_OUTPUT, SERVER_OUTPUT, HAS_SERVER } = CONFIG

const buildTarget = (target: BuildTarget, environment: BuildEnvironment = 'prod') => {
    const config = getWebpackConfig(target, environment)

    if (!config) {
        return Promise.reject(`Could not load webpack configuration for ${target}/${environment}!`)
    }

    return webpackPromise(config)
}

const cleanAndBuild = (target: BuildTarget, environment: BuildEnvironment = 'prod') => {
    const cleanTarget = target === 'server'
        ? SERVER_OUTPUT
        : CLIENT_OUTPUT

    return clean(cleanTarget).then(() => buildTarget(target, environment))
}

const getBuildEnvironment = (args: BuildParam[]) => {
    for (const arg of args) {
        if ((ENVIRONMENTS as string[]).indexOf(arg) !== -1) {
            return arg as BuildEnvironment
        }
    }
    return 'prod'
}

const getBuildTargets = (args: BuildParam[]) => {
    for (const arg of args) {
        if (!HAS_SERVER && arg === 'server') {
            continue
        }
        if ((TARGETS as string[]).indexOf(arg) !== -1) {
            return [arg] as BuildTarget[]
        }
    }

    const defaultTargets = HAS_SERVER
        ? ['server', 'client']
        : ['client']

    return defaultTargets as BuildTarget[]
}

/**
 * Builds one or all targets in a specified environment
 * @param args
 *  [complete] [<target>] [<environment>]
 *  - target defaults to both server and client
 *  - environment defaults to prod
 *  - complete: runs clean, lint and test before building
 */
const build = async (...args: BuildParam[]) => {
    const targets = getBuildTargets(args)
    const environment = getBuildEnvironment(args)

    if (args.indexOf('complete') !== -1) {
        await cleanBin()
        await lint()
        await test('--silent')
        return failPromisesLate(
            targets.map((target) => buildTarget(target, environment)),
        )
    } else {
        return failPromisesLate(
            targets.map((target) => cleanAndBuild(target, environment)),
        )
    }
}

export default build
