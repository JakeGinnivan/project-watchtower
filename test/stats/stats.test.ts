import fs from 'fs'
import path from 'path'
import clean from '../../lib/bin/clean'
import build from '../../lib/bin/build'
import stats from '../../lib/bin/stats'
import bundleSize from '../../lib/stats/bundle-size'
import ssrStats from '../../lib/stats/ssr-stats'
import lighthouseStats from '../../lib/stats/lighthouse'
import { getTestPort } from '../test-helpers'
import { getConfig, getRuntimeConfigFromBuildConfig } from '../../lib/runtime/config/config'
import { createConsoleLogger } from '../../lib/runtime/universal'

const log = createConsoleLogger()
const testProjectDir = path.join(process.cwd(), './test/test-project')
const buildConfig = getConfig(log, testProjectDir)
buildConfig.OUTPUT = path.resolve(buildConfig.BASE, 'test-dist/stats')

const runtimeConfig = getRuntimeConfigFromBuildConfig(buildConfig)

// Increase test timeout because builds might take a while
;(jasmine as any).DEFAULT_TIMEOUT_INTERVAL = 200000

describe('stats', () => {
    beforeAll(async () => {
        const port = await getTestPort()
        buildConfig.DEV_SERVER_PORT = port
        process.env.PROJECT_DIR = './test/test-project'

        await clean(log, buildConfig)
        await build(log, buildConfig)
    })

    it('bundle-size', async () => {
        const metrics = await bundleSize(log, runtimeConfig)

        const main = metrics.bundle_size_main
        const vendor = metrics.bundle_size_vendor
        const css = metrics.bundle_size_css

        expect(main).toBeDefined()
        expect(vendor).toBeDefined()
        expect(css).toBeDefined()

        expect(+main).not.toBeCloseTo(0)
        expect(+vendor).not.toBeCloseTo(0)
        expect(+css).not.toBeCloseTo(0)
    })

    it('ssr-stats', async () => {
        const metrics = await ssrStats(log, buildConfig)

        const documentSize = metrics.home_ssr_document_size
        const domSize = metrics.home_ssr_dom_size
        const loadTime = metrics.home_ssr_loadtime

        expect(documentSize).toBeDefined()
        expect(domSize).toBeDefined()
        expect(loadTime).toBeDefined()

        expect(+documentSize).not.toBeCloseTo(0)
        expect(+domSize).not.toBeCloseTo(0)
        expect(+loadTime).not.toBeCloseTo(0)
    })

    it('lighthouse', async () => {
        const metrics = await lighthouseStats(log, buildConfig)

        expect(metrics.home_first_meaningful_paint).toBeDefined()
        expect(metrics.home_speed_index).toBeDefined()
        expect(metrics.home_time_to_interactive).toBeDefined()
        expect(metrics.home_consistently_interactive).toBeDefined()
        expect(metrics.home_dom_size).toBeDefined()
        expect(metrics.home_perf_score).toBeDefined()
    })

    it('cli', async () => {
        await clean(log, buildConfig)
        await build(log, buildConfig)
        await stats(log, buildConfig, 'verbose')

        const filePath = path.resolve(testProjectDir, 'build-stats.csv')
        expect(fs.existsSync(filePath)).toBe(true)
    })
})
