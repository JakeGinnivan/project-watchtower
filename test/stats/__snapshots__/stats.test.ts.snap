// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats bundle-size 1`] = `
Object {
  "bundle_main.js": "replaced",
  "total_css": "replaced",
  "total_js": "replaced",
}
`;

exports[`stats lighthouse 1`] = `
Object {
  "home_bootup_time": "replaced",
  "home_category_accessibility_score": "replaced",
  "home_category_best-practices_score": "replaced",
  "home_category_performance_score": "replaced",
  "home_category_pwa_score": "replaced",
  "home_category_seo_score": "replaced",
  "home_estimated_input_latency": "replaced",
  "home_first_contentful_paint": "replaced",
  "home_first_cpu_idle": "replaced",
  "home_first_meaningful_paint": "replaced",
  "home_interactive": "replaced",
  "home_maximum_child_elements": "replaced",
  "home_maximum_dom_depth": "replaced",
  "home_network_requests": "replaced",
  "home_speed_index": "replaced",
  "home_total_byte_weight": "replaced",
  "home_total_dom_nodes": "replaced",
}
`;

exports[`stats ssr-stats 1`] = `
Object {
  "home_ssr_document_size": "replaced",
  "home_ssr_dom_size": "replaced",
  "home_ssr_loadtime": "replaced",
}
`;
